### PR TITLE
(#448) Clarify logic in task.c; Add unit-test to validate public interfaces.

### PR DESCRIPTION
**Pull request by gapisback**
This commit adds a bit of code cleanup to the task system module, task.c . No new functionality or change-in-logic is being done. Add new unit-test, task_system_test, to validate internal state of the task system for simple task creation / execution.

---
NOTE: to the reviewer. This change set came out of my code walk through to understand the task system and its innards.
This was done in context of the design rework to support execution of Splinter using allocated shared memory segments.
While at it, I constructed a small C-unit test to work through task system interfaces and to confirm, via tests, expected outputs.

Some of this testing will also be helpful when we "cut-over" to running tests using shared segments.

### CODE REVIEW

File: src/task.c [hunk 1]
```diff
@@ -10,23 +10,29 @@
 
 int              hook_init_done = 0;
 static int       num_hooks      = 0;
-static task_hook hooks[MAX_HOOKS];
+// RESOLVE: At 8 bytes per task_hook{}, we are burning up 16K of memory. Only 2 hooks
```
Comment (gapisback) [inline src/task.c:13]: @ajhconway - We should discuss and reconsider this. 

When we, eventually, move to a shared memory implementation, I suspect (expect) that this array will need to be allocated from shared memory segment. That will cost us 16K, whereas something like 64 bytes should suffice if we reduce MAX_HOOKS to 8.

File: src/task.c [hunk 2]
```diff
@@ -10,23 +10,29 @@
...
+// are currently being connected. What about reducing MAX_HOOKS to, say, 8?
+static task_hook hooks[MAX_HOOKS];	
 
 // forward declarations that aren't part of the public API of the task system
-platform_status
+static platform_status
```
Comment (gapisback) [inline src/task.c:18]: Many changes in this file add `static` to functions / prototypes to document clearly that this is a static function.
Improves code readability.

File: src/task.c [hunk 3]
```diff
@@ -83,12 +104,21 @@ task_get_max_tid(task_system *ts)
    return *max_tid + 1;
 }
 
+/*
+ * Return the bitmasks of tasks active. Mainly intended as a testing hook.
+ */
+uint64
+task_active_tasks_mask(task_system *ts)
```
Comment (gapisback) [inline src/task.c:111]: Added this extern accessor function, used in unit-tests for validating bitmask.

File: src/task.c [hunk 4]
```diff
@@ -149,7 +196,7 @@ task_register_this_thread(task_system *ts, uint64 scratch_size)
 void
 task_deregister_this_thread(task_system *ts)
 {
-   uint64 tid = platform_get_tid();
+   threadid tid = platform_get_tid();
```
Comment (gapisback) [inline src/task.c:199]: Use the right typedef.

File: src/task.c [hunk 5]
```diff
@@ -168,6 +215,17 @@ typedef struct {
...
 
+/*
+ * -----------------------------------------------------------------------------
+ * task_invoke_with_hooks() -
+ *
+ * A thread has been created with a call-back function to execute. This
+ * function invokes that function bracketted by calls to register / deregister
+ * the thread with Splinter's task system.
+ *
+ * RESOLVE: I see references to this, ... but who calls this?
```
Comment (gapisback) [inline src/task.c:226]: Let's discuss .... this, and then there seem to be a few more functions, that are written but never called.

Is this vestigial code that we could / should get rid of now, to reduce confusion ... [???].

File: src/task.c [hunk 6]
```diff
@@ -185,6 +243,11 @@ task_invoke_with_hooks(void *func_and_args)
    platform_free(thread_to_start->heap_id, func_and_args);
 }
 
+/*
+ * -----------------------------------------------------------------------------
+ * task_create_thread_with_hooks() - Comment ?
```
Comment (gapisback) [inline src/task.c:248]: Help! Can you help me add a meaningful prolog?

File: src/task.c [hunk 7]
```diff
@@ -215,6 +278,15 @@ task_create_thread_with_hooks(platform_thread       *thread,
    return ret;
 }
 
+/*
+ * -----------------------------------------------------------------------------
+ * task_thread_create() - Helper method to create a new task and to do the
+ *  required registration with Splinter.
+ *
+ * Currently, this is active mainly in tests. It's also used to create
+ * background tasks, a feature that is currently not enabled.
```
Comment (gapisback) [inline src/task.c:287]: @ajhconway - RESOLVE: What's the history on background threads?

Can I elicit some documentation from you and put it down here, so future readers can benefit?

File: src/task.c [hunk 8]
```diff
@@ -8,25 +8,32 @@
 
 #define MAX_HOOKS (2048)
 
-int              hook_init_done = 0;
-static int       num_hooks      = 0;
+int        hook_init_done = 0;
+static int num_hooks      = 0;
+// RESOLVE: At 8 bytes per task_hook{}, we are burning up 16K of memory. Only 2
+// hooks are currently being connected. What about reducing MAX_HOOKS to, say,
+// 8?
```
Comment (ajhconway) [inline src/task.c:15]: This does feel a little much. Maybe something still hard to imagine blowing out like 64?

File: src/task.c [hunk 9]
```diff
@@ -68,9 +84,15 @@ init_threadid(task_system *ts)
 
 out:
    debug_assert(tid != INVALID_TID);
+
+   // Sets thread-ID, tracked as thread-local storage.
```
Comment (ajhconway) [inline src/task.c:88]: This is only true on Linux. It could be implemented differently on another platform.

I would just rephrase this as "Sets thread-ID, for example tracked as thread-local storage."

File: src/task.c [hunk 10]
```diff
@@ -98,19 +129,30 @@ register_init_tid_hook(void)
...
    uint64 *tid_bitmask = task_system_get_tid_bitmask(ts);
-   platform_assert(!(*tid_bitmask & (1ULL << tid)));
+
+   uint64 bitmask_val = *tid_bitmask;
+
+   // Ensure that caller is only clearing for a thread that's in-use.
+   platform_assert(!(bitmask_val & (1ULL << tid)),
+                   "Thread [%lu] is expected to be in-use. Bitmap: 0x%lx",
+                   tid,
+                   bitmask_val);
```
Comment (ajhconway) [inline src/task.c:154]: This is good

File: tests/unit/task_system_test.c [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+   threadid main_thread_idx = platform_get_tid();
+   ASSERT_EQUAL(main_thread_idx, 0, "main_thread_idx=%lu", main_thread_idx);
+
+   // Setup thread-specific struct, needed for validation in thread's worker fn
+   thread_cfg.tasks                  = data->tasks;
+   thread_cfg.exp_thread_idx         = 1; // Main thread is at index 0.
+   thread_cfg.active_threads_bitmask = task_active_tasks_mask(data->tasks);
+
+   int rc = pthread_create(&new_thread, NULL, exec_one_thread, &thread_cfg);
+   ASSERT_EQUAL(0, rc);
```
Comment (ajhconway) [inline tests/unit/task_system_test.c:164]: This should be using `platform_thread` and `task_thread_create`.

File: tests/unit/task_system_test.c [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+      // These are independent of the new thread's creation.
+      thread_cfgp->tasks          = data->tasks;
+      thread_cfgp->exp_thread_idx = tctr;
+
+      rc =
+         pthread_create(&new_thread, NULL, exec_one_of_n_threads, thread_cfgp);
+      ASSERT_EQUAL(0, rc);
+
+      thread_cfgp->this_thread_id = new_thread;
+   }
```
Comment (ajhconway) [inline tests/unit/task_system_test.c:216]: Again this should be using `platform_thread` and `task_thread_create`.

File: tests/unit/task_system_test.c [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+   // Verify that the state of active-threads bitmask (managed by Splinter) has
+   // not changed just by creating this pthread. It should be the same as what
+   // we had recorded just prior to pthread_create().
+   ASSERT_EQUAL(thread_cfg->active_threads_bitmask,
+                task_bitmask_before_register);
+
+   // This is the important call to initialize thread-specific stuff in
+   // Splinter's task-system, which sets up the thread-id (index) and records
+   // this thread as active with the task system.
+   task_register_this_thread(thread_cfg->tasks, trunk_get_scratch_size());
```
Comment (ajhconway) [inline tests/unit/task_system_test.c:251]: Testing this is fine, but if we use the task thread creation machinery as I suggest above, then this call will happen as a hook.

This function is generally used only for the `main` thread or for threads otherwise created by a non-splinter context. So perhaps we should run this test on the main thread instead.

**Review by ajhconway — CHANGES_REQUESTED** [review]
Thank you for doing this. These comments, tests and fixes are all very good.

A couple of testing changes would be good though.
1. The tests shouldn't call the pthread functions, but should call their platform and task wrappers (see comments in code).
2. We don't test `task_enqueue`, `task_perform_one` and `task_perform_all`, which is the primary way that the task system is used in the core code. It's fine if this is out of scope for this PR though.

File: src/task.c (hunk 7 above)
Comment (gapisback) [reply src/task.c:288]: Thanks, @ajhconway - I will update the prolog accordingly.

Have opened issue #450 requesting that we have to work on re-enabling background threads support.

File: src/task.c (hunk 9 above)
Comment (gapisback) [reply src/task.c:88]: Done.

File: src/platform_linux/platform.c [hunk 14]
```diff
@@ -101,6 +101,9 @@ platform_buffer_destroy(buffer_handle *bh)
    return STATUS_OK;
 }
 
+/*
+ * platform_thread_create() - External interface to create a Splinter thread.
+ */
```
Comment (gapisback) [inline src/platform_linux/platform.c:106]: (Added many such prologs to make the whole module easier to follow.)

File: src/platform_linux/platform.c [hunk 15]
```diff
@@ -136,6 +139,12 @@ platform_thread_join(platform_thread thread)
    return CONST_STATUS(ret);
 }
 
+platform_thread
+platform_thread_id_self()
```
Comment (gapisback) [inline src/platform_linux/platform.c:143]: Added, externalized for testing.

File: src/task.c [hunk 16]
```diff
@@ -6,27 +6,31 @@
 
 #include "poison.h"
 
-#define MAX_HOOKS (2048)
+#define MAX_HOOKS (8)
```
Comment (gapisback) [inline src/task.c:9]: Reduced this .... Is this good enough or too low?  Let's discuss.

File: src/task.c [hunk 17]
```diff
@@ -40,8 +44,17 @@ task_init_tid_bitmask(uint64 *tid_bitmask)
...
+
+ * The first thing a newly started thread that wants to use Splinter is expected
+ * to do is to "register itself" with the task system. This function is
+ * defined as a thread's hook and gets run as part of the thread's
+ * registration. This function initializes each thread's thread-ID, which
+ * is really an index into the array of thread-specific data structures.
+ */
 static void
-init_threadid(task_system *ts)
+task_init_threadid(task_system *ts)
```
Comment (gapisback) [inline src/task.c:57]: Renamed with `task_` prefix to be consistent with other names in this file.

File: src/task.c [hunk 18]
```diff
@@ -133,23 +168,86 @@ task_register_hook(task_hook newhook)
...
+   /*
+    * For all newly created threads, the thread-index will be 0 before the hook
+    * function, run below, generates this thread's index. We do not want to
+    * register a thread twice; otherwise, it will simply burn up an index and
+    * possibly allocated scratch space.
+    * As threads generally will require scratch space to be allocated, an
+    * attempt to re-register a thread should find scratch space already
+    * allocated. Flag that.
+    */
+   platform_assert((thread_tid == 0)
```
Comment (gapisback) [inline src/task.c:195]: Added ... Useful assertion. Found and tripped on incorrect usages in test code that was resulting in nested register thread for the same task.

File: src/task.c [hunk 19]
```diff
@@ -133,23 +168,86 @@ task_register_hook(task_hook newhook)
...
+task_deregister_thread(task_system *ts,
+                       const char  *file,
+                       const int    lineno,
+                       const char  *func)
 {
-   uint64 tid = platform_get_tid();
+   threadid tid = platform_get_tid();
+
+   // A registered thread should always have a non-zero tid (index).
+   if (tid == 0) {
```
Comment (gapisback) [inline src/task.c:243]: Useful check ... Found bugs in test code that was doing 2 de-registers, inadvertently.

File: src/task.c [hunk 20]
```diff
@@ -168,24 +266,43 @@ typedef struct {
...
+ * the call-back function 'func' to now call Splinter APIs to do diff things.
+ * -----------------------------------------------------------------------------
+ */
 static void
 task_invoke_with_hooks(void *func_and_args)
 {
-   thread_invoke         *thread_to_start = (thread_invoke *)func_and_args;
-   platform_thread_worker func            = thread_to_start->func;
-   void                  *arg             = thread_to_start->arg;
+   thread_invoke         *thread_started = (thread_invoke *)func_and_args;
```
Comment (gapisback) [inline src/task.c:284]: Renamed variable from `thread_to_start` to `thread_started`, because by the time this function executes, it's already executing in the context of a newly fired up thread.

(Prior variable name was meaningful & correct in this fn's caller, where the thread was just going to be started.)

File: src/task.h [hunk 21]
```diff
@@ -127,12 +127,25 @@ task_thread_create(const char            *name,
                    platform_thread       *thread);
 
 // Register the calling thread, allocating scratch space for it
+#define task_register_this_thread(ts, scratch_size)                            \
+   task_register_thread((ts), (scratch_size), __FILE__, __LINE__, __FUNCTION__)
```
Comment (gapisback) [inline src/task.h:131]: Changed this and on L141 below to become caller-macros, so I can pass-down code location tags to the lower level function, for improved diagnostics.

File: tests/unit/task_system_test.c [hunk 22]
```diff
@@ -0,0 +1,496 @@
...
+ * in a real application code-flow). These interfaces will do all the stuff
+ * required to start a thread and set it up to perform Splinter work.
+ *
+ * This test case and its minion worker function,
+ * exec_one_thread_use_extern_apis(), is written to use external platform
+ * and task-system APIs. Otherwise, the testing done is identical to what's
+ * done in its sibling test case, test_one_thread_using_lower_apis.
+ * ------------------------------------------------------------------------
+ */
+CTEST2(task_system, test_one_thread_using_extern_apis)
```
Comment (gapisback) [inline tests/unit/task_system_test.c:224]: Added this new test case in 2nd commit, based on your review comments.

File: tests/unit/task_system_test.c (hunk 13 above)
Comment (gapisback) [reply tests/unit/task_system_test.c:343]: @ajhconway -- Not sure I understand what to change here.

With the revised set of tests, let me know if your comment above is still relevant and do I need to readjust the test code in any way

File: tests/unit/task_system_test.c [hunk 23]
```diff
@@ -0,0 +1,496 @@
...
+ * exec_one_thread_use_extern_apis() - Worker routine executed by a single
+ * thread.
+ *
+ * This worker method uses external interfaces to start a thread, thru
+ * task_invoke_with_hooks(). That interface registers the new thread with
+ * Splinter -before- invoking this worker function. Adjust the task-bitmask
+ * checks accordingly.
+ */
+static void
+exec_one_thread_use_extern_apis(void *arg)
```
Comment (gapisback) [inline tests/unit/task_system_test.c:410]: Worker fn added new in 2nd commit after your comments.

**Comment by gapisback** [comment]
HI, @ajhconway -- I have revised the code and test code based on your feedback. Have done semi-major overhaul of the test code.

I think this should cover most of your feedback. The change set is back into the review cycle now

Thanks!

**Comment by gapisback** [comment]
@ajhconway - I have addressed your comment (1), by reworking the test case and add a new one using just the external APIs. 

About (2): "_We don't test task_enqueue, task_perform_one and task_perform_all, which is the primary way that the task system is used in the core code. It's fine if this is out of scope for this PR though._"

I tried to work that in, but that will take me [more] time to understand how all these interfaces work and then to cobble together an execution flow which will invoke them. 

If you think that is very desirable, I can open an issue, but -- really -- I don't think I will have much time for that this month. (This work came out of Am behind my schedule to complete the PG-SplinterDB integration work, and am already running late on that.)

File: src/task.c (hunk 19 above)
Comment (ajhconway) [reply src/task.c:243]: Good catch

File: tests/unit/task_system_test.c (hunk 13 above)
Comment (ajhconway) [reply tests/unit/task_system_test.c:343]: Good now.

**Review by ajhconway — APPROVED** [review]
This looks good to me now, and I appreciate all the comments and cleanup.

One thing I would like is to preface the task tests with something to indicate that the external APIs are the preferred way of interacting and that the low level ones are not. I'm just trying to imagine a user looking at the test code as a template and trying to push them towards the external APIs.
